### PR TITLE
KAFKA-13113: Support unregistering Raft listeners

### PR DESCRIPTION
**Pull request by jsancio**
Support unregistering by returning a ListenerContext on registration and exposing a close method on the returned ListenerContext. To allow the user to use the same Listener on different registrations the associated ListenerContext is sent through all of the methods described by the Raft Listener.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: metadata/src/test/java/org/apache/kafka/metalog/LocalLogManager.java [hunk 1]
```diff
@@ -362,24 +363,27 @@ LeaderAndEpoch notifiedLeader() {
...
             this.offset = offset;
         }
 
         void beginShutdown() {
-            listener.beginShutdown();
+            listener.beginShutdown(this);
         }
+
+        @Override
+        public void close() {}
```
Comment (jsancio) [inline metadata/src/test/java/org/apache/kafka/metalog/LocalLogManager.java]: Should fix this by appending an event to eventQueue that removes this listener.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 2]
```diff
@@ -2491,6 +2492,47 @@ public void testLateRegisteredListenerCatchesUp() throws Exception {
...
+            .appendToLog(2, batch3)
+            .withUnknownLeader(epoch - 1)
+            .build();
+
+        context.becomeLeader();
+        context.client.poll();
+        assertEquals(10L, context.log.endOffset().offset);
+
+        // Let the initial listener catch up
+        context.deliverRequest(context.fetchRequest(epoch, otherNodeId, 10L, epoch, 0));
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: Use the helper method "advance high-watermark" in a few of these places.

**Comment by cmccabe** [comment]
I don't see why we need this context object. The listener itself is the context object. It should be able to unregister itself by just calling `RaftClient#unregister(this)`

You need a context object in non-object-oriented languages like C, where you can't call `Foo#bar(a, b)`, but you have to call `bar(Foo foo, a, b)` Since Java is object-oriented, the context is in `Foo` already, so no need for a helper object.

**Comment by jsancio** [comment]
> I don't see why we need this context object. The listener itself is the context object. It should be able to unregister itself by just calling `RaftClient#unregister(this)`
> 
> You need a context object in non-object-oriented languages like C, where you can't call `Foo#bar(a, b)`, but you have to call `bar(Foo foo, a, b)` Since Java is object-oriented, the context is in `Foo` already, so no need for a helper object.

Yeah. At a high-level we need to support two requirements:
1. Unregistering a Listener.
2. The Controller needs to know if a given "handleCommit", "handleSnapshot" or "handleLeaderChange" is for the previous registration or the current registration.

With the implementation in this PR the controller can do the following to determine if those events are for the current registration or the previous registration:
https://github.com/apache/kafka/pull/11116/files#diff-77dc2adb187fd078084644613cff2b53021c8a5fbcdcfa116515734609d1332a

If we don't expose the `ListenerContext` then the Controller needs to use a different `Listener` instance in the new registration so that it can determine if the "handleCommit", "handleSnapshot" or "handleLeaderChange" are for the current registration. With this other solution it is up to the user (Controller) to guarantee this and the RaftClient cannot help with this check.

I am okay with either solution as there are pros and cons to both.

**Comment by cmccabe** [comment]
Thanks for the explanation. I'd prefer just using a different listener instance. Should be straightforward for the controller + broker to do that since they use small helper objects for this anyway.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 3]
```diff
@@ -2381,25 +2396,51 @@ public void complete() {
...
+                }
+            } else {
+                if (contexts.remove(listener) == null) {
+                    logger.error("Attempting to remove a listener context that doesn't exists: {}", listenerName(listener));
+                }
+            }
+        }
+    }
+
+    private final class KafkaListenerContext implements CloseListener<BatchReader<T>> {
```
Comment (hachikuji) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: nit: is the `Kafka` prefix adding much value here?

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 4]
```diff
@@ -2381,25 +2396,51 @@ public void complete() {
...
+        private final Listener<T> listener;
+
+        private Registration(RegistrationOps ops, Listener<T> listener) {
+            this.ops = ops;
+            this.listener = listener;
+        }
+
+        private void update(Map<Listener<T>, KafkaListenerContext> contexts) {
+            if (ops == RegistrationOps.REGISTER) {
+                if (contexts.putIfAbsent(listener, new KafkaListenerContext(listener)) != null) {
```
Comment (hachikuji) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: Could we have an INFO message for the happy paths in here?

File: raft/src/main/java/org/apache/kafka/raft/RaftClient.java [hunk 5]
```diff
@@ -89,15 +89,32 @@ default void beginShutdown() {}
...
+     *
+     * @param listener the listener to register
      */
     void register(Listener<T> listener);
 
+    /**
+     * Unregisters a listener.
+     *
+     * To distinguish from events that happend before the call to {@code unregister} and a future
+     * call to {@code register}, then different {@code Listener} instances must be used.
```
Comment (hachikuji) [inline raft/src/main/java/org/apache/kafka/raft/RaftClient.java]: nit: remove "then"

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 6]
```diff
@@ -2381,25 +2396,51 @@ public void complete() {
...
+    private final class Registration {
+        private final RegistrationOps ops;
+        private final Listener<T> listener;
+
+        private Registration(RegistrationOps ops, Listener<T> listener) {
+            this.ops = ops;
+            this.listener = listener;
+        }
+
+        private void update(Map<Listener<T>, KafkaListenerContext> contexts) {
```
Comment (hachikuji) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: nit: this feels a tad unnatural. We already have an implicit reference to `listenerContexts`. Could we move this method into `KafkaRaftClient` and make `Registration` a static class?

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java (hunk 4 above)
Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: Done.

File: raft/src/main/java/org/apache/kafka/raft/RaftClient.java (hunk 5 above)
Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/RaftClient.java]: Done

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java (hunk 6 above)
Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: Done.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java (hunk 3 above)
Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: Done.

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the patch!

**Pull request merged**